rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+
+#### Closure-support For Functions
+
+For iteration to land in the form proposed in this RFC, we need a way for
+operators to write _what_ they want to happen to keys and/or values of objects
+and arrays.
+
+We do this by allowing functions to expose the fact that they accept a closure
+as a stand-alone argument to their function call.
+
+"stand-alone" means the closure comes _after_ the function call itself, e.g.
+this:
+
+```coffee
+map(.) |k, v| {
+  [k, v]
+}
+```
+
+over this:
+
+```coffee
+map(., |k, v| {
+  [k, v]
+})
+```
+
+This choice is made to make it clear that closures in VRL _can't be passed
+around through variables, but are instead syntactically attached to a function
+call_.
+
+That is, we don't want to allow this:
+
+```coffee
+my_closure = |k, v| [k, v]
+map(., my_closure)
+```
+
+As that would require us to (1) allow for a closure to be represented as a `Value`,
+since everything assigned to a variable has to be representable as a value, and
+(2) it would require us to add compiler-support to reject assigning closures to
+an event field (e.g. `.foo = |k, v| [k, v]`), as that would lead to its own set
+of problems.
+
+While closures-as-values might seem useful at first, there hasn't been any
+use-case that has come up in the past that makes this a requirement.
+
+Instead, the closure-syntax is tied to a function call, and can only be added to
+functions that explicitly expose their ability to take a closure with `x`
+arguments that returns `y` value.
+
+#### Lexical Scoping
+
+Lexical scoping (variables being accessible within a given scope, instead of
+globally) is something we've discussed before.
+
+Before, we decided that the complexity of adding lexical scoping wasn't worth
+the investment before our first release, and we also hoped that lexical scoping
+wouldn't be something that was ever needed in VRL.
+
+With this feature, and particular the function-closure syntax, lexical scoping
+comes to top of mind again.
+
+The reason for that, is the following example:
+
+```coffee
+map(.) |key, value| {
+  key = upcase(key)
+
+  [key, value]
+}
+
+key
+```
+
+We reference `key` outside the closure, at the last line of the program. What
+should the value of `key` be in this case?
+
+Without lexical scoping, it would be set to the upper-case variant of the "last"
+key in the event.
+
+With lexical scoping, it would return an "undefined variable" error at
+compile-time, because the `key` variable _inside_ the closure is
+lexically-scoped to that block, and remains undefined outside of the block.
+
+However, while the above syntax would be _new_ and thus not a breaking change,
+for existing code, adding lexical scoping _would_ be a breaking change:
+
+```coffee
+foo = "bar"
+
+{
+  foo = "baz"
+}
+
+foo
+```
+
+Previously, `foo` would return `"baz"` when the program runs, but with lexical
+scoping, it would be set to `"bar"` instead (depending on if we also added
+let-bindings or not).
```

Comment:
A minor detail, but I don't think this is quite right. Scoping is more about where the variables are declared and available rather than what happens to the previous values.

The confusion is that in VRL we don't have the concept of declaring variables, so we bring them into scope the first type a value is set.

I would argue that in the previous example, the `foo = "bar"` statement declares the variable, then when we enter the block that definition is moved into the scope of the block and `foo = "baz"` just sets the value of the original `foo`. So with or without lexical scoping in that example `foo` should return "baz".

Where it would break would be if you had:

```coffee
{
  foo = "baz"
}
foo
```

In this case, `foo` would not exist outside the block, currently it would and would return "baz".

Reply:
Agreed. I do think without variable assignment syntax (e.g. let-bindings), you can't do certain things you might want to do in terms of assigning/re-assigning, but I don't think we _need_ to introduce that in this RFC, if at all.

I'll update the RFC to match what you noted here.